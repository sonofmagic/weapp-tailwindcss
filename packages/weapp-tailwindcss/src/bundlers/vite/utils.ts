import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
