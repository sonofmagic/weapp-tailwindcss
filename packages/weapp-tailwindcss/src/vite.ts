export * from './bundlers/vite/index'
