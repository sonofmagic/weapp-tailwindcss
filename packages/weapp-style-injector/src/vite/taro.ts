import type { TaroSubPackageConfig } from '../taro'
import type { ViteWeappStyleInjectorOptions } from '../vite'

import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { createTaroSubPackageImportResolver } from '../taro'
import { mergePerFileResolvers, toArray } from '../utils'
import weappStyleInjector from '../vite'

export interface ViteTaroStyleInjectorOptions extends Omit<ViteWeappStyleInjectorOptions, 'perFileImports'> {
  appConfigPath?: string | string[]
  subPackages?: TaroSubPackageConfig | TaroSubPackageConfig[]
  perFileImports?: ViteWeappStyleInjectorOptions['perFileImports']
}

function resolveDefaultAppConfigPaths(): string[] {
  const cwd = process.cwd()
  return [
    path.resolve(cwd, 'src/app.config.ts'),
    path.resolve(cwd, 'src/app.config.js'),
    path.resolve(cwd, 'src/app.config.json'),
    path.resolve(cwd, 'app.config.ts'),
    path.resolve(cwd, 'app.config.js'),
    path.resolve(cwd, 'app.config.json'),
  ]
}

export function weappStyleInjectorTaro(options: ViteTaroStyleInjectorOptions = {}) {
  const {
    appConfigPath,
    subPackages,
    perFileImports,
    ...rest
  } = options

  const configs = new Map<string, TaroSubPackageConfig>()

  for (const entry of toArray(subPackages)) {
    configs.set(path.resolve(entry.appConfigPath), entry)
  }

  const candidatePaths = appConfigPath
    ? toArray(appConfigPath).map(entry => path.resolve(entry))
    : resolveDefaultAppConfigPaths()

  for (const candidate of candidatePaths) {
    if (!configs.has(candidate) && fs.existsSync(candidate)) {
      configs.set(candidate, { appConfigPath: candidate })
    }
  }

  const taroResolver = createTaroSubPackageImportResolver(Array.from(configs.values()))

  return weappStyleInjector({
    ...rest,
    perFileImports: mergePerFileResolvers([perFileImports, taroResolver]),
  })
}

export default weappStyleInjectorTaro
