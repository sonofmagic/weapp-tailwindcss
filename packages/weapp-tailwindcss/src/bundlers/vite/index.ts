import type { Plugin } from 'vite'
import type { OutputAsset, OutputChunk } from 'rollup'
import type { UserDefinedOptions } from '@/types'
import { getOptions } from '@/options'
import { vitePluginName } from '@/constants'
import { getGroupedEntries } from '@/utils'
import { createTailwindcssPatcher } from '@/tailwindcss/patcher'
import { createDebug } from '@/debug'

const debug = createDebug()

/**
 * @name UnifiedViteWeappTailwindcssPlugin
 * @description uni-app vite vue3 版本插件
 * @link https://weapp-tw.icebreaker.top/docs/quick-start/frameworks/uni-app-vite
 */
export function UnifiedViteWeappTailwindcssPlugin(options: UserDefinedOptions = {}): Plugin | undefined {
  const opts = getOptions(options)
  const {
    disabled,
    onEnd,
    onLoad,
    onStart,
    onUpdate,
    templateHandler,
    styleHandler,
    patch,
    jsHandler,
    mainCssChunkMatcher,
    appType,
    setMangleRuntimeSet,
    cache,
    tailwindcssBasedir,
  } = opts
  if (disabled) {
    return
  }

  patch?.()

  const twPatcher = createTailwindcssPatcher()
  onLoad()
  // 要在 vite:css 处理之前运行
  return {
    name: vitePluginName,
    enforce: 'post',
    async generateBundle(opt, bundle) {
      debug('start')
      onStart()

      const entries = Object.entries(bundle)
      const groupedEntries = getGroupedEntries(entries, opts)
      const runtimeSet = twPatcher.getClassSet({
        basedir: tailwindcssBasedir,
      })
      setMangleRuntimeSet(runtimeSet)
      debug('get runtimeSet, class count: %d', runtimeSet.size)
      if (Array.isArray(groupedEntries.html)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.html.length; i++) {
          const [file, originalSource] = groupedEntries.html[i] as [string, OutputAsset]

          const oldVal = originalSource.source.toString()

          const hash = cache.computeHash(oldVal)
          cache.calcHashValueChanged(file, hash)

          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.source = source
                debug('html cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              originalSource.source = await templateHandler(oldVal, {
                runtimeSet,
              })
              onUpdate(file, oldVal, originalSource.source)
              debug('html handle: %s', file)
              noCachedCount++
              return {
                key: file,
                source: originalSource.source,
              }
            },
          )
        }
        debug('html handle finish, total: %d, no-cached: %d', groupedEntries.html.length, noCachedCount)
      }
      if (Array.isArray(groupedEntries.js)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.js.length; i++) {
          const [file, originalSource] = groupedEntries.js[i] as [string, OutputChunk]
          const rawSource = originalSource.code

          const hash = cache.computeHash(rawSource)
          cache.calcHashValueChanged(file, hash)
          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.code = source
                debug('js cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              const mapFilename = `${file}.map`
              const hasMap = Boolean(bundle[mapFilename])
              const { code, map } = await jsHandler(rawSource, runtimeSet, {
                generateMap: hasMap,
              })
              originalSource.code = code
              onUpdate(file, rawSource, code)
              debug('js handle: %s', file)
              noCachedCount++
              if (hasMap && map) {
                ;(bundle[mapFilename] as OutputAsset).source = map.toString()
              }
              return {
                key: file,
                source: code,
              }
            },
          )
        }
        debug('js handle finish, total: %d, no-cached: %d', groupedEntries.js.length, noCachedCount)
      }

      if (Array.isArray(groupedEntries.css)) {
        let noCachedCount = 0
        for (let i = 0; i < groupedEntries.css.length; i++) {
          const [file, originalSource] = groupedEntries.css[i] as [string, OutputAsset]

          const rawSource = originalSource.source.toString()

          const hash = cache.computeHash(rawSource)
          cache.calcHashValueChanged(file, hash)
          await cache.process(
            file,
            () => {
              const source = cache.get<string>(file)
              if (source) {
                originalSource.source = source
                debug('css cache hit: %s', file)
              }
              else {
                return false
              }
            },
            async () => {
              const css = await styleHandler(rawSource, {
                isMainChunk: mainCssChunkMatcher(originalSource.fileName, appType),
              })
              originalSource.source = css
              onUpdate(file, rawSource, css)
              debug('css handle: %s', file)
              noCachedCount++
              return {
                key: file,
                source: css,
              }
            },
          )
        }
        debug('css handle finish, total: %d, no-cached: %d', groupedEntries.css.length, noCachedCount)
      }

      onEnd()
      debug('end')
    },
  }
}
