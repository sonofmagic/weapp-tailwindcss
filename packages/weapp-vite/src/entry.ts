import process from 'node:process'
import mm from 'micromatch'
import { defu } from '@weapp-core/shared'
import path from 'pathe'
import klaw from 'klaw'
import { defaultExcluded, getWxmlEntry, searchAppEntry } from './utils'
import type { Entry, InlineConfig } from './types'

export function createFilter(include: string[], exclude: string[], options?: mm.Options) {
  const opts = defu<mm.Options, mm.Options[]>(options, {
    ignore: exclude,
    // dot: true,
    // contains: true,
  })
  return function (id: unknown | string) {
    if (typeof id !== 'string') {
      return false
    }
    if (/\0/.test(id)) {
      return false
    }

    return mm.isMatch(id as string, include, opts)
  }
}

export async function getEntries(options: { root?: string, srcRoot?: string, outDir?: string, relative?: boolean, subPackage?: InlineConfig['subPackage'] }) {
  // build.outDir
  const { root = process.cwd(), outDir = 'dist', relative, srcRoot = '', subPackage } = options

  function formatPath(to: string) {
    if (relative) {
      return path.relative(root, to)
    }
    return path.normalize(to)
  }

  // 单独打包分包的场景
  if (subPackage) {
    const subPackageRoot = subPackage.root ?? ''
    const filter = createFilter(
      [path.join(subPackageRoot, '**/*')],
      [
        ...defaultExcluded,
        path.join(root, `${outDir}/**`),
      ],
      { cwd: root },
    )

    const pageEntries: Entry[] = []
    const componentEntries: Entry[] = []

    for await (
      const file of klaw(
        path.join(root, subPackageRoot),
        {
          filter,
        },
      )
    ) {
      if (file.stats.isFile()) {
        if (/\.wxml$/.test(file.path)) {
          const entry = getWxmlEntry(file.path, formatPath)
          if (entry) {
            //  && !walkPathsSet.has(file.path)
            // 防止重复，理论上不会
            // walkPathsSet.add(file.path)
            if (entry.type === 'component') {
              componentEntries.push(entry)
            }
            else if (entry.type === 'page') {
              pageEntries.push(entry)
            }
          }
        }
      }
    }
    return {
      pages: pageEntries,
      components: componentEntries,
    }
  }
  // 打包主包的场景
  else {
    const appEntry = searchAppEntry(root, {
      formatPath,
    })

    // TODO exclude 需要和 output 配套
    // const walkPathsSet = new Set<string>()
    if (appEntry) {
      const subPackageDeps = appEntry.deps.filter(x => x.type === 'subPackage')
      const filter = createFilter(
        [path.join(srcRoot, '**/*')],
        [
          ...defaultExcluded,
          path.join(root, `${outDir}/**`),
          ...subPackageDeps.map((x) => {
            return path.join(root, `${x.root}/**`)
          }),
        ],
        { cwd: root },
      )

      const pageEntries: Entry[] = []
      const componentEntries: Entry[] = []

      for await (
        const file of klaw(
          path.join(root, srcRoot),
          {
            filter,
          },
        )
      ) {
        if (file.stats.isFile()) {
          if (/\.wxml$/.test(file.path)) {
            const entry = getWxmlEntry(file.path, formatPath)
            if (entry) {
              //  && !walkPathsSet.has(file.path)
              // 防止重复，理论上不会
              // walkPathsSet.add(file.path)
              if (entry.type === 'component') {
                componentEntries.push(entry)
              }
              else if (entry.type === 'page') {
                pageEntries.push(entry)
              }
            }
          }
        }
      }

      return {
        app: appEntry,
        pages: pageEntries,
        components: componentEntries,
        subPackages: subPackageDeps,
        // walkPathsSet,
      }
    }
  }
}
