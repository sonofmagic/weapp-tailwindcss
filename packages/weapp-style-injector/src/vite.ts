import type { Plugin } from 'vite'
import type { WeappStyleInjectorOptions } from './core'
import type { UniAppManualStyleConfig, UniAppSubPackageConfig } from './uni-app'
import { createStyleInjector, PLUGIN_NAME } from './core'
import { createUniAppSubPackageImportResolver } from './uni-app'
import { mergePerFileResolvers } from './utils'

export interface ViteWeappStyleInjectorOptions extends WeappStyleInjectorOptions {
  uniAppSubPackages?: UniAppSubPackageConfig | UniAppSubPackageConfig[]
  uniAppStyleScopes?: UniAppManualStyleConfig | UniAppManualStyleConfig[]
}

export function weappStyleInjector(options: ViteWeappStyleInjectorOptions = {}): Plugin {
  const {
    uniAppSubPackages,
    uniAppStyleScopes,
    perFileImports,
    ...restOptions
  } = options

  const perFileResolver = mergePerFileResolvers([
    typeof perFileImports === 'function' ? perFileImports : undefined,
    createUniAppSubPackageImportResolver(uniAppSubPackages, uniAppStyleScopes),
  ])

  const injector = createStyleInjector({
    ...restOptions,
    perFileImports: perFileResolver,
  })

  return {
    name: PLUGIN_NAME,
    apply: 'build',
    enforce: 'post',
    async generateBundle(_, bundle) {
      if (!injector.hasImports) {
        return
      }

      for (const [fileName, output] of Object.entries(bundle)) {
        if (output.type !== 'asset') {
          continue
        }
        if (!injector.shouldProcess(fileName)) {
          continue
        }

        const source = typeof output.source === 'undefined' ? '' : output.source
        const result = injector.inject(fileName, source)

        if (!result.changed) {
          continue
        }

        output.source = result.content
      }
    },
  }
}

export default weappStyleInjector
