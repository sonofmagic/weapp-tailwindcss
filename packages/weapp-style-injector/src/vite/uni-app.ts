import type { OutputAsset, OutputBundle, OutputOptions, PluginContext } from 'rollup'
import type { Plugin, PluginContainer, ResolvedConfig } from 'vite'
import type { ResolvedSubPackage, UniAppStyleScopeInput, UniAppSubPackageConfig } from '../uni-app'

import type { ViteWeappStyleInjectorOptions } from '../vite'
import fs from 'node:fs'

import path from 'node:path'
import process from 'node:process'
import { preprocessCSS } from 'vite'
import { resolveUniAppStyleScopes, splitUniAppStyleScopes } from '../uni-app'
import { toArray } from '../utils'
import weappStyleInjector from '../vite'

export interface ViteUniAppStyleInjectorOptions extends Omit<ViteWeappStyleInjectorOptions, 'uniAppSubPackages'> {
  pagesJsonPath?: string | string[]
  subPackages?: UniAppSubPackageConfig | UniAppSubPackageConfig[]
  indexFileName?: string | string[]
  styleScopes?: UniAppStyleScopeInput | UniAppStyleScopeInput[]
}

function resolveDefaultPagesJsonPaths(): string[] {
  const cwd = process.cwd()
  return [
    path.resolve(cwd, 'src/pages.json'),
    path.resolve(cwd, 'pages.json'),
  ]
}

// 使用 Tailwind 插件的 transform 钩子时需要兼容多种写法，这里先抽象出统一的处理函数。
type TransformHandler = (this: PluginContext, code: string, id: string) => unknown

function getTransformHandler(hook: Plugin['transform'] | undefined): TransformHandler | undefined {
  if (!hook) {
    return undefined
  }

  if (typeof hook === 'function') {
    return hook as TransformHandler
  }

  if (typeof hook === 'object' && hook !== null && 'handler' in hook && typeof hook.handler === 'function') {
    return hook.handler as TransformHandler
  }

  return undefined
}

// 构建阶段生成分包入口样式时，提前触发 Tailwind 的 transform 逻辑并缓存结果
function createUniAppSubPackageIndexEmitter(subPackages: ResolvedSubPackage[]): Plugin {
  const existing = [...subPackages]
  if (existing.length === 0) {
    return {
      name: 'weapp-style-injector:uni-app-sub-packages',
      apply: 'build' as const,
    }
  }

  let resolvedConfig: ResolvedConfig | undefined
  let pluginContainer: PluginContainer | undefined
  let tailwindTransformHandler: TransformHandler | undefined
  const processedSourceCache = new Map<string, string>()
  const outputCache = new Map<string, string>()

  const transformContext = {
    addWatchFile() {},
  } as unknown as PluginContext

  // Vite 配置里 plugins 可能是多层嵌套数组，统一拍平成线性列表便于检索
  const flattenPlugins = (plugins: ReadonlyArray<Plugin | Plugin[] | null | undefined | false>): Plugin[] => {
    const flat: Plugin[] = []
    for (const entry of plugins) {
      if (Array.isArray(entry)) {
        flat.push(...flattenPlugins(entry))
      }
      else if (entry) {
        flat.push(entry)
      }
    }
    return flat
  }

  return {
    name: 'weapp-style-injector:uni-app-sub-packages',
    apply: 'build' as const,
    configResolved(config: ResolvedConfig) {
      resolvedConfig = config
      // Vite 在 build 阶段不会暴露 pluginContainer，这里通过类型断言尝试读取，失败时会退回到手动调用 Tailwind transform
      pluginContainer = (config as ResolvedConfig & { pluginContainer?: PluginContainer }).pluginContainer
      if (!tailwindTransformHandler) {
        const tailwindCandidates = flattenPlugins(config.plugins)
          .filter(entry => entry?.name === '@tailwindcss/vite:generate:build')
        const matched = tailwindCandidates.find(entry => typeof entry?.transform !== 'undefined')
        if (matched?.transform) {
          tailwindTransformHandler = getTransformHandler(matched.transform)
        }
      }
    },
    async generateBundle(_: OutputOptions, bundle: OutputBundle) {
      for (const entry of existing) {
        const sourcePath = entry.sourceAbsolutePath
        if (!fs.existsSync(sourcePath)) {
          continue
        }

        const fileName = entry.indexRelativePath
        let processedSource = outputCache.get(fileName)

        if (typeof processedSource === 'undefined') {
          const cacheKey = `${sourcePath}::${entry.preprocess !== false ? '1' : '0'}`
          processedSource = processedSourceCache.get(cacheKey)

          if (typeof processedSource === 'undefined') {
            let rawSource: string
            try {
              rawSource = await fs.promises.readFile(sourcePath, 'utf8')
            }
            catch {
              continue
            }

            if (entry.preprocess !== false && resolvedConfig) {
              try {
                let transformedSource = rawSource

                const initialResult = await preprocessCSS(transformedSource, sourcePath, resolvedConfig)
                transformedSource = initialResult.code

                const transformIds = new Set<string>([sourcePath, `${sourcePath}?direct`])

                if (pluginContainer) {
                  try {
                    const resolvedId = await pluginContainer.resolveId(sourcePath)
                    if (resolvedId?.id) {
                      transformIds.add(resolvedId.id)
                      transformIds.add(`${resolvedId.id}?direct`)
                    }
                  }
                  catch {
                    // ignore resolution errors and fall back to raw path
                  }
                }

                for (const id of transformIds) {
                  let transformResult: unknown

                  if (pluginContainer) {
                    transformResult = await pluginContainer.transform(transformedSource, id)
                  }
                  else if (tailwindTransformHandler) {
                    // 当无法获取 pluginContainer 时，直接调用 Tailwind 插件的 transform 以确保 @tailwind 指令被展开
                    transformResult = await tailwindTransformHandler.call(transformContext, transformedSource, id)
                  }

                  if (
                    transformResult
                    && typeof transformResult === 'object'
                    && 'code' in transformResult
                    && typeof (transformResult as { code?: unknown }).code === 'string'
                  ) {
                    transformedSource = (transformResult as { code: string }).code
                    break
                  }
                }

                const finalResult = await preprocessCSS(transformedSource, sourcePath, resolvedConfig)
                processedSource = finalResult.code
              }
              catch (error) {
                throw Object.assign(
                  new Error(`[weapp-style-injector] Failed to preprocess "${sourcePath}": ${(error as Error).message}`),
                  { cause: error },
                )
              }
            }
            else {
              processedSource = rawSource
            }

            processedSourceCache.set(cacheKey, processedSource)
          }

          outputCache.set(fileName, processedSource)
        }

        const existingAsset = bundle[fileName]
        if (existingAsset && existingAsset.type === 'asset') {
          existingAsset.source = processedSource
          continue
        }

        bundle[fileName] = {
          type: 'asset',
          name: fileName,
          fileName,
          source: processedSource,
        } as OutputAsset
      }
    },
  }
}

export function StyleInjector(options: ViteUniAppStyleInjectorOptions = {}) {
  const {
    pagesJsonPath,
    subPackages,
    indexFileName,
    styleScopes,
    ...rest
  } = options

  const configs = new Map<string, UniAppSubPackageConfig>()
  const { subPackages: scopedSubPackages, manual: manualStyleScopes } = splitUniAppStyleScopes(styleScopes)

  for (const entry of [...toArray(subPackages), ...scopedSubPackages]) {
    configs.set(path.resolve(entry.pagesJsonPath), entry)
  }

  const candidatePaths = pagesJsonPath
    ? toArray(pagesJsonPath).map(entry => path.resolve(entry))
    : resolveDefaultPagesJsonPaths()

  for (const candidate of candidatePaths) {
    if (!configs.has(candidate) && fs.existsSync(candidate)) {
      configs.set(candidate, {
        pagesJsonPath: candidate,
        indexFileName,
      })
    }
  }

  const entries = configs.size > 0 ? Array.from(configs.values()) : undefined
  const manualEntries = manualStyleScopes.length > 0 ? manualStyleScopes : undefined
  const resolvedSubPackages = resolveUniAppStyleScopes(entries, manualEntries)

  const plugins = [
    weappStyleInjector({
      ...rest,
      uniAppSubPackages: entries,
      uniAppStyleScopes: manualEntries,
    }),
  ]

  if (resolvedSubPackages.length > 0) {
    plugins.unshift(createUniAppSubPackageIndexEmitter(resolvedSubPackages))
  }

  return plugins.length === 1 ? plugins[0] : plugins
}
