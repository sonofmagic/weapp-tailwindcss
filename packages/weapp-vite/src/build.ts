import type { InlineConfig } from 'vite'
import { build } from 'vite'
import type { RollupOutput, RollupWatcher } from 'rollup'
import { defu } from '@weapp-core/shared'
import { vitePluginWeapp } from './plugins'
import { getEntries } from './entry'

export function getDefaultConfig(options: { cwd?: string, entries?: string[] }): InlineConfig {
  return {
    build: {
      rollupOptions: {
        output: {
          format: 'cjs',
          entryFileNames: (chunkInfo) => {
            return chunkInfo.name
          },
        },
      },
      assetsDir: '.',
      commonjsOptions: {
        transformMixedEsModules: true,
        include: undefined,
      },
    },
    plugins: [
      vitePluginWeapp(options),
    ],
  }
}

export async function runDev(cwd: string, options?: InlineConfig) {
  const entries = await getEntries(cwd)

  if (entries) {
    const watcher = (await build(
      defu<InlineConfig, InlineConfig[]>(
        options,
        getDefaultConfig({
          cwd,
          entries: entries.all,
        }),
        {
          build: {
            watch: {},
            minify: false,
          },
        },
      )
      ,
    )) as RollupWatcher

    // watcher.on('event', (event) => {
    //   console.log(event)
    // })

    return watcher
  }
}

export async function runProd(cwd: string, options?: InlineConfig) {
  const entries = await getEntries(cwd)

  if (entries) {
    const output = (await build(
      defu<InlineConfig, InlineConfig[]>(
        options,
        getDefaultConfig({
          cwd,
          entries: entries.all,
        }),
      ),
    )) as RollupOutput | RollupOutput[]

    return output
  }
}
