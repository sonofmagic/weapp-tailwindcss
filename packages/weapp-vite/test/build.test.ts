import path from 'pathe'
import { build } from 'vite'
import type { RollupOutput } from 'rollup'

import { runProd } from '@/build'
import { getEntries } from '@/entry'

describe('build', () => {
  describe('rollup', () => {
    const mixjsDir = path.resolve(__dirname, './fixtures/mixjs')
    // function r(...args: string[]) {
    //   return path.resolve(mixjsDir, ...args)
    // }

    it('mixjs getEntries', async () => {
      const entries = await getEntries({
        cwd: mixjsDir,
        relative: true,
      })
      expect(entries).toMatchSnapshot()
    })

    it('mixjs vite build', async () => {
      function relative(p: string) {
        return path.relative(mixjsDir, p)
      }
      const entries = await getEntries({
        cwd: mixjsDir,
      })
      if (!entries) {
        return
      }
      const input = entries.all
        .reduce<Record<string, string>>((acc, cur) => {
          acc[relative(cur)] = cur
          return acc
        }, {})

      const res = await build({
        build: {
          rollupOptions: {
            input,
            output: {
              format: 'cjs',
              entryFileNames: (chunkInfo) => {
                return chunkInfo.name
              },
            },
          },
          assetsDir: '.',
          write: false,
          commonjsOptions: {
            include: undefined,
            transformMixedEsModules: true,
          },
          minify: false,
        },
        // plugins: [
        //   {
        //     name: 'inspect',
        //     enforce: 'pre',
        //     configResolved(_config) {
        //       // console.log(config)
        //     },
        //     resolveId(source) {
        //       console.log('resolveId', source)
        //     },
        //     load(id, options) {
        //       console.log('load', id, options)
        //     },
        //     generateBundle(_options, _bundle) {
        //       console.log('generateBundle', _bundle)
        //     },
        //   },
        // ],

      }) as RollupOutput

      for (const item of res.output) {
        // @ts-ignore
        expect(item.code).toMatchSnapshot(path.normalize(item.fileName))
      }
    })

    it('mixjs runProd', async () => {
      const res = await runProd(mixjsDir, {
        build: {
          minify: false,
          commonjsOptions: {
            include: undefined,
            transformMixedEsModules: true,
          },
        },

        plugins: [
          {
            name: 'inspect',
            enforce: 'pre',
            configResolved(_config) {
              // console.log(config)
            },
            resolveId(source) {
              console.log('resolveId', source)
            },
            load(id, options) {
              console.log('load', id, options)
            },
            generateBundle(_options, _bundle) {
              console.log('generateBundle', _bundle)
            },
            buildEnd(error) {
              console.log(error)
            },
          },
        ],
      }) as RollupOutput

      for (const item of res.output) {
        // @ts-ignore
        expect(item.code).toMatchSnapshot(path.normalize(item.fileName))
      }
    })
  })
})
