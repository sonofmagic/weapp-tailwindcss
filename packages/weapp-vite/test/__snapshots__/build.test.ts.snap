// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`build > rollup > mixjs getEntries 1`] = `
{
  "app": {
    "deps": [
      {
        "path": "pages/index",
        "type": "page",
      },
    ],
    "json": {
      "pages": [
        "pages/index",
      ],
    },
    "jsonPath": "app.json",
    "path": "app.js",
    "type": "app",
  },
  "components": [
    {
      "deps": [],
      "json": {
        "component": true,
      },
      "jsonPath": "index2.json",
      "path": "index2.js",
      "type": "component",
    },
  ],
  "pages": [
    {
      "deps": [],
      "path": "index.js",
      "type": "page",
    },
    {
      "deps": [],
      "path": "pages/index.js",
      "type": "page",
    },
  ],
  "subPackages": [],
}
`;

exports[`build > rollup > mixjs runProd abs root > app.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const x = cjs.cjs;
console.log(x);
"
`;

exports[`build > rollup > mixjs runProd abs root > app.json 1`] = `
"{
  "pages": [
    "pages/index"
  ]
}"
`;

exports[`build > rollup > mixjs runProd abs root > cjs-B3a6j7Rd.js 1`] = `
""use strict";
const a = "1";
var cjs = {
  a,
  b: () => {
    return "b";
  }
};
exports.cjs = cjs;
"
`;

exports[`build > rollup > mixjs runProd abs root > esm-CR02SQDf.js 1`] = `
""use strict";
const c = "c";
exports.c = c;
"
`;

exports[`build > rollup > mixjs runProd abs root > index.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs runProd abs root > index.wxml 1`] = `"<view class="text-gray-50">index.wxml</view>"`;

exports[`build > rollup > mixjs runProd abs root > index2.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs runProd abs root > index2.json 1`] = `
"{
  "component": true
}"
`;

exports[`build > rollup > mixjs runProd abs root > index2.wxml 1`] = `"<view>Index2</view>"`;

exports[`build > rollup > mixjs runProd abs root > pages/index.js 1`] = `
""use strict";
console.log("pages/index.wxml");
"
`;

exports[`build > rollup > mixjs runProd abs root > pages/index.wxml 1`] = `"<view class="text-gray-50">pages/index.wxml</view>"`;

exports[`build > rollup > mixjs vite build > app.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const x = cjs.cjs;
console.log(x);
"
`;

exports[`build > rollup > mixjs vite build > cjs-B3a6j7Rd.js 1`] = `
""use strict";
const a = "1";
var cjs = {
  a,
  b: () => {
    return "b";
  }
};
exports.cjs = cjs;
"
`;

exports[`build > rollup > mixjs vite build > esm-CR02SQDf.js 1`] = `
""use strict";
const c = "c";
exports.c = c;
"
`;

exports[`build > rollup > mixjs vite build > index.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs vite build > index2.js 1`] = `
""use strict";
const cjs = require("./cjs-B3a6j7Rd.js");
const esm = require("./esm-CR02SQDf.js");
const a = cjs.cjs;
console.log(a, esm.c);
"
`;

exports[`build > rollup > mixjs vite build > pages/index.js 1`] = `
""use strict";
console.log("pages/index.wxml");
"
`;
